terraform/eval_error.go

```diff
@@ -0,0 +1,16 @@
+package terraform
+
+// EvalReturnError is an EvalNode implementation that returns an
+// error if it is present.
+//
+// This is useful for scenarios where an error has been captured by
+// another EvalNode (like EvalApply) for special EvalTree-based error
+// handling, and that handling has completed, so the error should be
+// returned normally.
+type EvalReturnError struct {
+	Error *error
+}
+
+func (n *EvalReturnError) Eval(ctx EvalContext) (interface{}, error) {
+	return nil, *n.Error
```

Comment:
I would change this to also handle the case where `n.Error` is a nil pointer. At the moment you'll get a panic dereferencing a nil pointer.